pkg/endpoint/endpoint.go

```diff
@@ -197,36 +203,60 @@ func RunK8sCiliumEndpointSyncGC() {
 					}
 				}
 
-				clusterPodSet := map[string]bool{}
-				clusterPods, err := k8sClient.CoreV1().Pods("").List(meta_v1.ListOptions{})
-				if err != nil {
-					return err
-				}
-				for _, pod := range clusterPods.Items {
-					podFullName := pod.Name + ":" + pod.Namespace
-					clusterPodSet[podFullName] = true
-				}
+				var (
+					listOpts = meta_v1.ListOptions{Limit: 10}
+					loopStop = time.Now().Add(CiliumEndpointGCInterval)
+				)
 
-				// "" is all-namespaces
-				ceps, err := ciliumClient.CiliumEndpoints(meta_v1.NamespaceAll).List(meta_v1.ListOptions{})
-				if err != nil {
-					scopedLog.WithError(err).Debug("Cannot list CEPs")
-					return err
-				}
-				for _, cep := range ceps.Items {
-					cepFullName := cep.Name + ":" + cep.Namespace
-					if _, found := clusterPodSet[cepFullName]; !found {
-						// delete
-						scopedLog = scopedLog.WithFields(logrus.Fields{
-							logfields.EndpointID: cep.Status.ID,
-							logfields.K8sPodName: cepFullName,
-						})
-						scopedLog.Debug("Orphaned CiliumEndpoint is being garbage collected")
-						if err := ciliumClient.CiliumEndpoints(cep.Namespace).Delete(cep.Name, &meta_v1.DeleteOptions{}); err != nil {
-							scopedLog.WithError(err).Debug("Unable to delete CEP")
+			perCEPFetch:
+				for time.Now().Before(loopStop) { // stop the loop if it takes too long
```

Comment:
What can take too long?

Reply:
It makes sure that a single controller run does not exceed the interval period

Reply:
This is really guarding against an infinite loop. I picked the interval period since it was a well-defined time. I'm mostly worried about not calling `break` somewhere in the code.